packages/buidler-etherscan/src/solc/bytecode.ts

```diff
@@ -0,0 +1,264 @@
+import { RunTaskFunction } from "@nomiclabs/buidler/types";
+
+import { metadataLengthSize, readSolcMetadataLength } from "./metadata";
+import { InferralType } from "./version";
+
+export async function lookupMatchingBytecode(
+  contractFiles: CompilerOutput["contracts"],
+  deployedBytecode: string,
+  inferralType: InferralType
+) {
+  for (const [contractFilename, contracts] of Object.entries(contractFiles)) {
+    for (const [contractName, contract] of Object.entries(contracts)) {
+      // Normalize deployed bytecode according to this contract.
+      const { deployedBytecode: runtimeBytecodeSymbols } = contract.evm;
+
+      const comparison = await compareBytecode(
+        deployedBytecode,
+        runtimeBytecodeSymbols,
+        inferralType
+      );
+
+      if (comparison.match) {
+        const {
+          contractInformation: {
+            immutableValues,
+            libraryLinks,
+            normalizedBytecode,
+          },
+        } = comparison;
+        // The bytecode matches
+        return {
+          immutableValues,
+          libraryLinks,
+          normalizedBytecode,
+          contractFilename,
+          contractName,
+          contract,
+        };
+      }
+    }
+  }
+
+  return null;
+}
+
+type BytecodeComparison =
+  | { match: false }
+  | { match: true; contractInformation: BytecodeExtractedData };
+
+interface BytecodeExtractedData {
+  immutableValues: ImmutableValues;
+  libraryLinks: ResolvedLinks;
+  normalizedBytecode: string;
+}
+
+export async function compareBytecode(
+  deployedBytecode: string,
+  runtimeBytecodeSymbols: CompilerOutputBytecode,
+  inferralType: InferralType
+): Promise<BytecodeComparison> {
+  let bytecodeSize = deployedBytecode.length;
+  // We will ignore metadata information when comparing. Etherscan seems to do the same.
```

Comment:
What happens if we don't ignore the metadata? Is this block (and passing `inferralType` around) worth it?

Reply:
If we don't ignore the metadata and instead only accept identical metadata, we need to communicate properly causes for differing metadata. E.g. suggest configuring the `settings.metadata.bytecodeHash` property for the compiler input. I think that's doable but it might add another layer of complexity in the long run.

Reply:
Ugh, I see. I'm inclined to just let that happen and assume that most users won't face that, but it would be very annoying for those who do.